### PR TITLE
Fuzzing Coverage Expansion

### PR DESCRIPTION
**Pull request by viktoriia-lsg**
This pull requests expands fuzzing coverage  for OSS-fuzz by adding 4 more fuzz test files and 5 corpora files. All fuzz tests are written in native go fuzzing style. Corpora files are compatible with OSS-fuzz format.

### CODE REVIEW

File: huff0/huff0_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		_, remain, err := ReadTable(buf0, &s)
+		if err != nil {
+			return
+		}
+		out, err := s.Decompress1X(remain)
+		if err != nil || out == nil {
+			return
+		}
+	})
+}
```
Comment (klauspost) [inline huff0/huff0_fuzz_test.go:89]: Add newline, but I think `go fmt` will fix that.

File: s2/s2_fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package s2
```
Comment (klauspost) [inline s2/s2_fuzz_test.go:1]: Please merge into `fuzz_test.go`. To be honest this should just be in `FuzzEncodingBlocks`.

Adding a Decoding fuzzer would be of much more value.

File: s2/s2_fuzz_test.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+func FuzzS2(f *testing.F) {
+	fuzz.AddFromZip(f, "testdata/fuzz/block-corpus-raw.zip", fuzz.TypeRaw, false)
+	f.Fuzz(func(t *testing.T, data []byte) {
+		concat, err := ConcatBlocks(nil, data, []byte{0})
+		if err != nil || concat == nil {
+			return
+		}
+
+		if len(data) < 9 { //to avoid errors when converting to uint64
```
Comment (klauspost) [inline s2/s2_fuzz_test.go:16]: What is this working around? Smells like it is working around a bug.

File: s2/s2_fuzz_test.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+		}
+
+		encodeBlockGo(encoded, data)
+		encodeBlockBetterGo(encoded, data)
+		encodeBlockSnappyGo(encoded, data)
+		encodeBlockBetterSnappyGo(encoded, data)
+		dst := encodeGo(encoded, data)
+		if dst == nil {
+			return
+		}
```
Comment (klauspost) [inline s2/s2_fuzz_test.go:32]: So this "just" tests the encoders similar to `FuzzEncodingBlocks`, but avoids the assembly. I tend to just use tags for disabling assembly.

But if you want to add them, having them in that function makes sense to me.

File: snappy/xerial/xerial_encode_fuzz_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package xerial
```
Comment (klauspost) [inline snappy/xerial/xerial_encode_fuzz_test.go:1]: Please merge into `fuzz_test.go` - and TBH I don't really see where this adds value. Are you just removing the sub-tests?

File: huff0/huff0_fuzz_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
+package huff0
```
Comment (klauspost) [inline huff0/huff0_fuzz_test.go:1]: `fuzz_test.go`

**Review by klauspost — CHANGES_REQUESTED** [review]

**Comment by klauspost** [comment]
Please see #289 for reference.

A) I would like to move seed files off-repo, since it is making the repo too large. I have an [old repo](https://github.com/klauspost/compress-fuzz) that was used for the old fuzz tests, but can be repurposed for this.

B) I do not want to end up like with #289 - where a third party controls scripts that I cannot modify freely. This means I have been unable to modify/add my fuzz tests for half a year now. I have asked @AdamKorcz for assistance in this, but no response. If I am not in control of the setup, I will not accept this.

If you need assistance in making CI pass, let me know.

File: fse/fse_fuzz_test.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	f.Fuzz(func(t *testing.T, buf0 []byte) {
+		var s, s2 Scratch
+		b, err := Compress(buf0, &s)
+		if err != nil || b == nil {
+			return
+		}
+		err = s.validateNorm()
+		if err != nil {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (viktoriia-lsg) [reply fse/fse_fuzz_test.go:21]: @klauspost Hi! We've used "validateNorm" to increase code coverage. We can remove it from the fuzz test, if it's only used for debugging.

File: s2/s2_fuzz_test.go (hunk 3 above)
Comment (viktoriia-lsg) [reply s2/s2_fuzz_test.go:16]: I want to discuss this one before making changes to this file. During the local testing, I observed an error in "EstimateBlockSize" function. When providing input as "\x01\x00\x00\x00\x00\x00\x00\x00", the error occurs in load64 function from encode_all.go file when calling Uint64 method. It occurs in golang standard libraries and crashes the fuzzer.
I may propose to leave this workaround as it is or remove "EstimateBlockSize" function from the fuzz test.

Comment (klauspost) [reply s2/s2_fuzz_test.go:16]: That seems like a real problem. I will send a fix.

Comment (klauspost) [reply s2/s2_fuzz_test.go:16]: Fix sent in https://github.com/klauspost/compress/pull/867

When tests pass I will merge - you can then marge master and remove the check 👍🏼

Comment (klauspost) [reply s2/s2_fuzz_test.go:16]: Merged. You should be able to remove the check now.

Comment (klauspost) [reply s2/s2_fuzz_test.go:16]: Merged.

File: s2/s2_fuzz_test.go (hunk 4 above)
Comment (viktoriia-lsg) [reply s2/s2_fuzz_test.go:32]: I'll move them to FuzzEncodingBlocks function later. Could you please clarify, what you mean by "but avoids the assembly" and "I tend to just use tags for disabling assembly."

I've included these encode* functions to cover them during fuzzing.

File: snappy/xerial/xerial_encode_fuzz_test.go (hunk 5 above)
Comment (klauspost) [reply snappy/xerial/xerial_encode_fuzz_test.go:1]: Yeah. Just replace the existing.

File: fse/fse_fuzz_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (klauspost) [reply fse/fse_fuzz_test.go]: It is fine.

**Comment by klauspost** [comment]
Seems a file is missing:
```
--- FAIL: FuzzEncoder (0.00s)
    helpers.go:36: open testdata/fuzz/block-corpus-raw.zip: no such file or directory
FAIL
```

**Comment by viktoriia-lsg** [comment]
Thanks! I've added a missing file and slightly changed a path to it.

**Comment by klauspost** [comment]
```
Run diff <(gofmt -d .) <(printf "")
1,13d0
< diff fse/fuzz_test.go.orig fse/fuzz_test.go
< --- fse/fuzz_test.go.orig
< +++ fse/fuzz_test.go
< @@ -3,[8](https://github.com/klauspost/compress/actions/runs/6313933219/job/17143098604?pr=866#step:4:9) +3,8 @@
<  import (
<  	"bytes"
<  	"fmt"
< -	"testing"
<  	"github.com/klauspost/compress/internal/fuzz"
< +	"testing"
<  )
<
```

**Comment by klauspost** [comment]
> 1. Does it mean that fse/fuzz_test.go needs a newline after "testing" package?

Imports are always sorted.

> 2. Regarding s2_fuzz_test.go file, is everything fine, so I can merge it with s2/fuzz_test.go? [Fuzzing Coverage Expansion #866 (comment)](https://github.com/klauspost/compress/pull/866#discussion_r1337195032)

Yes, please do that.

> 3. What do you think about this setup? [Fuzzing Coverage Expansion #866 (comment)](https://github.com/klauspost/compress/pull/866#issuecomment-1735485344)

Sounds like a reasonable approach.

**Comment by viktoriia-lsg** [comment]
Alright, we started working on it.

**Comment by viktoriia-lsg** [comment]
Hi @klauspost! We've made changes as discussed earlier. We've also opened a pull request to compress-fuzz with seed corpuses https://github.com/klauspost/compress-fuzz/pull/2.

Please let us know, if we need to change anything.

**Comment by klauspost** [comment]
Thanks! I will run the tests and give it a final review.

**Review by klauspost — CHANGES_REQUESTED** [review]
Just one change and a question.

File: ossfuzz/ossfuzz.sh [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+go mod init setupdicts
+go mod tidy
+go run main.go --dict-path=$SRC/compress/zstd/testdata/dict-tests-small.zip --output-file=$SRC/compress/zstd/fuzzDicts.go
+cp $SRC/compress/zstd/fuzzDicts.go $OUT/
+# Done creating "dicts" variable.
+
+cd $SRC/compress
+
+# Temporarily use a fork of go-fuzz-headers with some improvements that has not been merged yet.
+go mod edit -replace github.com/AdaLogics/go-fuzz-headers=github.com/AdamKorcz/go-fuzz-headers-1@22e92b7968997eabd210694dd4825dd0d19b697c
```
[the remark this answers is not in the document]
Comment (viktoriia-lsg) [reply ossfuzz/ossfuzz.sh:37]: That's a temporary solution for go-fuzz-headers package, which used primarily by oss-fuzz to support golang fuzzing. I see that the specific commit was done to "github.com/AdaLogics/go-fuzz-headers" several months ago, so it should work fine without this line now. I'll test it to ensure that this line can be removed.

[the remark this answers is not in the document]
Comment (viktoriia-lsg) [reply ossfuzz/ossfuzz.sh]: It works fine without this line, so removed.

**Review by klauspost — APPROVED** [review]

**Comment by klauspost** [comment]
Looks good!

**Comment by viktoriia-lsg** [comment]
Thanks! Please also review https://github.com/klauspost/compress-fuzz/pull/2 , so I can make a PR to oss-fuzz in order to start fuzzing.

**Comment by klauspost** [comment]
> Thanks! Please also review [klauspost/compress-fuzz#2](https://github.com/klauspost/compress-fuzz/pull/2) , so I can make a PR to oss-fuzz in order to start fuzzing.

Ah. Somehow I missed that. Will take a look - probably tomorrow.

**Pull request merged**